### DISCUSSION TITLE
Booting Jetson-NX: Error opening sdcard-0

### DISCUSSION THREAD

**Discussion by fre-wolf** [OE4T/meta-tegra#1097] [Q&A] [answered]
Hi,
I have the Jetson Xavier NX emmc version on a custom baseboard with a microSD card reader. 
I managed to build and flash a linux onto the emmc and could, by with some new files, e.g. linux-tegra_4.9.bbappend, fully access a microSD card.

My plan is now to get the root file system onto that microSD card and boot from there (ideally similar to what the boot would be on a jetson-xavier-nx-devkit with microSD card).

I started by creating a new machine.conf following the guide in https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE and tried adjusting the build by using some lines of the non-emmc jetson-xavier-nx-devkit.conf, such that my file now looks like this:
```
require conf/machine/jetson-xavier-nx-devkit-emmc.conf
MACHINEOVERRIDES = "cuda:tegra:tegra194:xavier-nx:jetson-xavier-nx-devkit-emmc:${MACHINE}"
PACKAGE_EXTRA_ARCHS_append = " jetson-xavier-nx-devkit-emmc"

# from jetson-xavier-nx-devkit.conf:
EMMC_SIZE = ""
PARTITION_LAYOUT_TEMPLATE = "flash_l4t_t194_spi_sd_p3668.xml"
TEGRA_SPIFLASH_BOOT = "1"
TEGRAFLASH_SDCARD_SIZE = "16G"
```

The following ./doflash.sh resulted in a system on the emmc. I then used ./dosdcard.sh to generate an image for the sdcard. While booting, I saw this:
```
...
[0002.386] E> tegrabl_cbo_parse_info: "boot-configuration" not found in CBO file.
[0002.393] I> Using default boot order
[0002.397] I> boot-dev-order :-
[0002.400] I> 1.sd
[0002.401] I> 2.usb
[0002.403] I> 3.nvme
[0002.405] I> 4.emmc
[0002.407] I> 5.net
[0002.409] I> Hit any key to stop autoboot:     4       3       2

E> tegrabl_display_printf: display is not initialized

TEGRA194 # boot sd
I> found decompressor handler: lz4-legacy
I> decompressing BMP blob ...
I> Kernel type = Normal
I> ########## SD (0) boot ##########
I> Found sdcard
I> enabling 'vdd-sdmmc1-sw' regulator
I> regulator 'vdd-sdmmc1-sw' already enabled
W> Error opening sdcard-0
I> sdcard-0 params source = DT-BL
W> Failed to initialize device 6-0
E> SD boot failed, err: 2
I> ########## USB (0) boot ##########
W> No valid slot number is found in scratch register
W> Return default slot: _a
...
```
The system then continued booting from emmc.

My conclusion from that is, that I don't need to edit the boot order as the system tries to boot from SD already (How to do that, see #890 ).

1) What else do I have to change to boot from external SD-card? 
2) Do I have to adjust the following lines from conf/machine/include/xavier-nx.inc into some SD-card related parameters by also including them into my machine.conf? If no, what change do I have to make to my machine.conf? :
```
OTABOOTDEV ?= "/dev/mtdblock0"
OTAGPTDEV ?= "/dev/mtdblock0"
```
3) Is there a way to flash the external SD card and the spi with ./doflash.sh?

**Comment by madisongh** [comment]
If the SDcard is working from Linux but not from cboot, you probably need to break out the cboot sources to find out why.  It looks like it's finding the device tree entry for the device (the 'DT-BL' mentioned in the messages), but the driver in cboot isn't able to initialize it. Unfortunately, the 'Failed to initialize' error message isn't particularly helpful, so you may need to insert additional print statements to debug further.

You should not change the OTABOOTDEV and OTAGPTDEV settings - all Xavier NX SoMs use that SPI flash device for storing the boot firmware.

As far as using doflash.sh to flash the SDcard - I'm not sure.  You could try using the `flash_l4t_t194_spi_sd_p3668.xml` layout file to tell the flasher app firmware to load the content onto the SDcard, but if cboot isn't able to access the device, then I suspect that the flasher app won't be able to, either.

**Reply by fre-wolf** [reply]
Is it possible to set the sdcard to sd-instance=2 in machine.conf or related files or is my only chance changing cboot sources?

**Reply by madisongh** [reply]
If you're talking about the `sd_params` that get stashed in the MB1 BCT, you could add a bbappend for the tegra-bootfiles recipe that modifies the `tegra194-mb1-bct-misc-sd-l4t.cfg` with the settings you need.

**Comment by fre-wolf** [comment, marked as answer]
Solved now. 
Other than the new machine.conf based on jetson-xavier-nx-devkit with the SKU and other parameters from the eMMC version, I had to edit following files:
in recipe **cboot-t19x:**
linux_load.c: copied case BOOT_FROM_USB to BOOT_FROM_SD and set device_instance in the SD-case to 2 (was 0)
config_storage.c: set sd_instance to 2 (was 0)
in recipe **linux-tegra:**
tegra194-p3668-common.dtsi: added a node for sdhci_sd1: sdhci@3440000 (copied from provided files by board vendor). In addition, I disabled all other sd and emmc related nodes:
```
sdhci_sd1: sdhci@3440000 {
        status = "okay";
        keep-power-in-suspend;
        /delete-property/ non-removable;
        /delete-property/ mmc-ddr-1_8v;
        mmc-ocr-mask = <0>;
        sd-uhs-sdr104;
        sd-uhs-sdr50;
        sd-uhs-sdr25;
        sd-uhs-sdr12;
        uhs-mask = <0x8>;
        max-clk-limit = <208000000>;
        tap-delay = <3>;
        cd-inverted;
        nvidia,cd-wakeup-capable;
        vmmc-supply = <&p3668_vdd_sdmmc1_sw>;
        nvidia,qvmmc-always-on;
        cd-gpios = <&tegra_main_gpio TEGRA194_MAIN_GPIO(Q, 2) 0>;
    };
```
tegra194-power-tree-p3668.dtsi: added a node that is specifying the powersupply for the node sdhci@3440000.
```
sdhci@3440000 {
        vmmc-supply = <&p3668_vdd_sdmmc1_sw>;
    };
```
